web3.js/src/connection.ts

```diff
@@ -4592,15 +4686,30 @@ export class Connection {
   async removeSlotUpdateListener(
     clientSubscriptionId: ClientSubscriptionId,
   ): Promise<void> {
-    if (this._slotUpdateSubscriptions[clientSubscriptionId]) {
-      const subInfo = this._slotUpdateSubscriptions[clientSubscriptionId];
-      delete this._slotUpdateSubscriptions[clientSubscriptionId];
-      await this._unsubscribe(subInfo, 'slotsUpdatesUnsubscribe');
-      this._updateSubscriptions();
+    await this._unsubscribeClientSubscription(
+      clientSubscriptionId,
+      'slot update',
+    );
+  }
+
+  /**
+   * @internal
+   */
+
+  async _unsubscribeClientSubscription(
+    clientSubscriptionId: ClientSubscriptionId,
+    subscriptionName: string,
+  ) {
+    const dispose =
+      this._subscriptionDisposeFunctionsByClientSubscriptionId[
+        clientSubscriptionId
+      ];
+    if (dispose) {
```

Comment:
How about deleting the dispose function from `_subscriptionDisposeFunctionsByClientSubscriptionId` within this `if` block? Then the dispose function doesn't have to worry about whether it has been called before. The responsibility for that check will belong to this `_unsubscribeClientSubscription` function.

Reply:
Yeah, sorry, this is just muscle memory from years of damage. Because functions can be passed around like values, it becomes inevitable that _someday_ someone will find a way to get a reference to the actual dispose function itself. Once that's done, they can call it over and over and over, no matter that we've removed it from the map.

In this particular case, we don't leak the dispose handles to callers, so what you've suggested will work. Will change!

Reply:
* I still want to leave the responsibility for deleting the dispose handle up to the dispose handle itself. A little self-destruct.

Reply:
Haha yeah good to be defensive I suppose, your change works for me.